### DISCUSSION TITLE
Add validation function for model fields

### DISCUSSION THREAD

**Discussion by Atlinx** [twinlogix/typetta#220] [Ideas]
To make security integration easier and more intuitive, Typetta should let you add validation logic on a per-field basis.

As mentioned in #214, changing the behaviour of the security policies to define a range of accepted values may be too specific. A more general implementation would be to let users provide their own validation functions for the fields they need it in.

**EXAMPLE**

```typescript
// Used alone in unsecured EntityManager
new EntityManger({
  validate: {
    userRole: {
      roleCode: (newValue) => {
        return securityContextPerms.MANAGE_ROLES;
      },
    },
    userSocial: {
      name: (newValue) => {
        return newValue != "github";
      }
    }
  }
});
```

```typescript
// Used in conjunction with a secured EntityManager, possibly using permissions within the validation
const securityContextPerms = ...

new EntityManager<never, { securityDomain: OperationSecurityDomain }, Permission, SecurityDomain>({
  validate: {
    userRole: {
      roleCode: (newValue) => {
        return securityContextPerms.MANAGE_ROLES;
      },
    },
    userSocial: {
      name: (newValue) => {
        return newValue != "github";
      }
    }
  },
  security: {
    context: {
      permissions: securityContextPerms ?? {}
    },
    policies: {
      user: {
        domain: {
          userId: "id"
        },
	permission: {
          UPDATE_PROFILE: {
            read: true,
            write: {
              avatarLink: true,
              bannerLink: true,
              createdAt: true,
              id: true,
              name: true,
              projectMembers: true,
              roles: true,
              socials: true,
          
              email: false,
              loginIdentities: false,
            }
          }
        }
      }
    }
  }
});
```

**Comment by edobrb** [comment]
An interesting concept, however I don't understand the semantics: the validation of `roles` and `socials` fields to which entity do they refer? The check is done before insertion and update?
What if my field validation logic depends on another field of the same entity?
A similar discussion is #157

**Reply by Atlinx** [reply]
I was thinking the validation happens before insertion and update. If your logic depends on other fields, you will have to fetch those fields using an EntityManager before you can apply your validation logic.

**Reply by Atlinx** [reply]
Whoops I realized I didn't specify the entity the fields -- I've updated my post to fix that.

**Reply by edobrb** [reply]
All clear. This feature is simply implementable through a middleware. I understand that having it as a first class feature directly on the EntityManager would be much more convenient. I'll tag this issue in #157

**Reply by Atlinx** [reply]
Now that I think about it, we could probably make this as an addon (maybe in a separate package?) that's a middleware. It could look something like:

```typescript
new EntityManager<never, { securityDomain: OperationSecurityDomain }, Permission, SecurityDomain>({
  middleware: [
    new ValidationMiddleware<...>({
          userRole: {
            roleCode: (newValue) => {
              return securityContextPerms.MANAGE_ROLES;
            },
          },
          userSocial: {
            name: (newValue) => {
              return newValue != "github";
            }
          }
    });
  ]
});
```

**Reply by Atlinx** [reply]
Is security implemented as a middleware right now? Maybe that could also be decoupled to a separate middleware as well to lighten up Typetta's core.

**Reply by edobrb** [reply]
Yes, exactly. FYI also the security is a middleware but in order to improve the user experience we also give the opportunity to specify all the needed parameters as first class under `security` field of the EntityManager.
